mapper/orm/src/main/java/org/hibernate/search/mapper/orm/massindexing/impl/IdentifierConsumerDocumentProducer.java

```diff
@@ -196,39 +197,85 @@ private void indexAllQueue(SessionImplementor session, List<?> entities) throws
 			return;
 		}
 
-		// TODO use 'SessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();' instead
-		PojoWorkPlan workPlan = mapping.createSearchManager( session ).createWorkPlan();
+		// TODO <<monitor>> indexingMonitor.entitiesLoaded( entities.size() );
+		PojoSessionWorkExecutor workExecutor = mapping.createSearchManager( session ).createSessionWorkExecutor();
+		CompletableFuture<?>[] tasks = new CompletableFuture<?>[entities.size()];
 
-		log.tracef( "received a list of objects to index: %s", entities );
-		for ( Object object : entities ) {
-			try {
-				index( workPlan, object );
-				// TODO: implements monitor
-				// monitor.documentsBuilt( 1 );
-			}
-			catch (RuntimeException e) {
-				String errorMsg = log.massIndexerUnableToIndexInstance(
-						object.getClass().getName(),
-						object.toString()
-				);
-
-				// TODO: implements exception handler
-				// errorHandler.handleException( errorMsg, e );
-				// temporary re-throw the exception
-				throw new RuntimeException( errorMsg, e );
+		for ( int i = 0; i < entities.size(); i++ ) {
+			tasks[i] = index( workExecutor, entities.get( i ) );
+		}
+
+		try {
+			CompletableFuture.allOf( tasks ).join();
+		}
+		catch (RuntimeException e) {
```

Comment:
This is not how you're supposed to use completable futures, especially not when you intend to (mostly) ignore and log exceptions.

A better implementation would be to add something like this, just after the `tasks[i] = index( workExecutor, entities.get( i ) );` (inside the loop):

```
tasks[i] = index( workExecutor, entities.get( i ) );
tasks[i].exceptionally( exception -> handleException( entities.get( i ), exception ) );
```


Then you can safely ignore the exception when you join, since it's handled on a per-work basis:

```
CompletableFuture.allOf( tasks ).exceptionally( exception -> null ).join();
```

Ideally we should use an instances of nested classes instead of lambdas, but you get the idea.

Reply:
Much better as you pointed out, thanks!